decompiler/pipeline/ssa/variable_renaming.py

```diff
@@ -334,3 +338,93 @@ def _classes_of(self, neighborhood: Iterable[Variable]) -> Iterable[Variable]:
+        """
+        self._color_classes is a dictionary where the set of keys is the set of colors
+        and to each color we assign the set of variables of this color.
+        """
+        super().__init__(task, interference_graph.copy())
+
+        dependency_graph = dependency_graph_from_cfg(task.graph)
+
+        mapping = {}
+        for variable in self.interference_graph.nodes():
+            contracted = tuple(self._variables_contracted_to[variable])
+            for var in contracted:
+                mapping[(var,)] = contracted
+
+        # Merge nodes which need to be contracted from self._variables_contracted_to
+        dependency_graph = networkx.relabel_nodes(dependency_graph, mapping)
+
+        # counter = 0
+        # self._decorate_graph(dependency_graph, f"dep{counter}.svg")
+
+        dependency_graph.edge = dependency_graph.edges(data=True)
+        while True:
+            for u, v, _ in sorted(dependency_graph.edges(data=True), key=lambda edge: edge[2]["score"], reverse=True):
+                if u == v:  # self loop
+                    continue
+
+                variables = u + v
+                if interference_graph.are_interfering(*variables):
+                    continue
+                if u[0].type != v[0].type:
+                    continue
+                if u[0].is_aliased != v[0].is_aliased:
+                    continue
+
+                break
+            else:
+                # We didn't find any remaining nodes to contract, break outer loop
+                break
+
+            networkx.relabel_nodes(dependency_graph, {u: (*u, *v), v: (*u, *v)}, copy=False)
+            # counter += 1
+            # self._decorate_graph(dependency_graph, f"dep{counter}.svg")
+
+        # counter += 1
+        # self._decorate_graph(dependency_graph, f"dep{counter}.svg")
+
+        self._variable_classes_handler = VariableClassesHandler(defaultdict(set))
+        for i, vars in enumerate(dependency_graph.nodes):
+            for var in vars:
+                self._variable_classes_handler.add_variable_to_class(var, i)
+
+        self.compute_new_name_for_each_variable()
+
+    def _decorate_graph(self, dependency_graph: MultiDiGraph, path: str):
+        decorated_graph = MultiDiGraph()
+        for node in dependency_graph.nodes:
+            decorated_graph.add_node(hash(node), label="\n".join(map(lambda n: f"{n}: {n.type}, aliased: {n.is_aliased}", node)))
+        for u, v, data in dependency_graph.edges.data():
+            decorated_graph.add_edge(u, v, label=f"{data['score']}")
+        for nodes in networkx.weakly_connected_components(dependency_graph):
+            for node_1, node_2 in combinations(nodes, 2):
+                if any(self.interference_graph.has_edge(pair[0], pair[1]) for pair in itertools.product(node_1, node_2)):
+                    decorated_graph.add_edge(hash(node_1), hash(node_2), color="red", dir="none")
+
+        DecoratedGraph(decorated_graph).export_plot(path, type="svg")
```

Comment:
This is debug stuff, I guess. If you think this is also helpful for other stages, you could add it to the decoration.py file which contains all the printing-functions for graphs

Reply:
The function only makes sense for the dependency graph which is currently only used by the conditional renaming. However, I moved this method to `dependency_graph.py` which I think is a better place for it. bb16b0b9